libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+        GCD a b p
+
+
+Uninhabited (Factor Z p) where
+    uninhabited (CofactorExists {n = Z} _ ok _) = uninhabited ok
+
+Uninhabited (Factor n Z) where
+    uninhabited (CofactorExists q ok prf) =
+        absurd . succNotLTEzero $ replace {P = LTE 1} (sym prf) ok
+
+||| Given a statement that p is factor of n, return its cofactor.
+cofactor : Factor n p -> (q : Nat ** Factor n q)
+cofactor (CofactorExists {n} {p} q ok prf) =
+        (q ** CofactorExists p ok $ rewrite multCommutative q p in prf)
+
+
+||| No number can simultaneously be and not be a factor of another number.
+factorNotFactorAbsurd : Factor n p -> NotFactor n p -> Void
+factorNotFactorAbsurd (CofactorExists {n} {p} q positN prf) (ProperRemExists {n} {p} q' r _ contra) =
+        thisIsAbsurd q q' prf $ replace {P = \x => (p * q') + S (finToNat r) = x} (sym prf) contra
+    where
+    thisIsAbsurd : (q, q' : Nat) -> p * q = n -> (p * q') + S (finToNat r) = p * q -> Void
+    thisIsAbsurd q q' nIsPQ a with (cmp q q')
+        thisIsAbsurd q (q + S d) nIsPQ a  | CmpLT d =
+            plusSuccIsNotIdentity .
+            replace {P = \x => (p * q) + x = p * q} (sym $ plusSuccRightSucc (p * S d) (finToNat r)) .
+            replace {P = \x => x = p * q} (sym $ plusAssociative (p * q) (p * S d) (S $ finToNat r)) $
+            replace {P = \x => x + S (finToNat r) = p * q} (multDistributesOverPlusRight p q (S d)) a
+        thisIsAbsurd q q nIsPQ a          | CmpEQ = plusSuccIsNotIdentity a
+        thisIsAbsurd (q + S d) q nIsPQ a  | CmpGT d =
+            let defSr =
+                    replace {P = \x => S (finToNat r) = x} (multRightSuccPlus p d) .
+                    subtractEqLeft $
+                    replace {P = \x => (p * q) + S (finToNat r) = x} (multDistributesOverPlusRight p q (S d)) a
+                (_ ** nIsSucc) = lteToSucc positN
+                pGtZ = nonZeroLeftFactor $ replace {P = \x => p * (q + S d) = x} nIsSucc nIsPQ
+            in
+            succNotLTEzero . subtractLteLeft .
+            replace {P = \x => LTE (p + S (p * d)) x} (sym $ plusZeroRightNeutral p) .
+            replace {P = \x => LTE x p} (plusSuccRightSucc p (p * d)) .
+            replace {P = \x => LTE (S x) p} defSr .
+            replace {P = LTE (S (S (finToNat r)))} (succPred p {ok = pGtZ}) $
+            LTESucc $ elemSmallerThanBound r
+
+||| The relation of common factor is symmetric, that is if p is a common factor
+||| of n and m, then it is also a common factor if m and n.
+commonFactorSym : CommonFactor a b p -> CommonFactor b a p
+commonFactorSym (CommonFactorExists p pfa pfb) = CommonFactorExists p pfb pfa
+
+||| 1 is a factor of any natural number.
+oneIsFactor : (n : Nat) -> {auto ok : LTE 1 n} -> Factor n 1
+oneIsFactor (S k) {ok} =
+        CofactorExists (S k) ok (rewrite plusZeroRightNeutral k in Refl)
+
+||| For all natural numbers n, n is (the greatest) a factor of n.
+selfFactor : (n : Nat) -> {auto ok : LTE 1 n} -> Factor n n
```

Comment:
I would call this `factorReflexive`. The factor relationship is a [partial order](https://github.com/idris-lang/Idris-dev/blob/0417c53fb6b91fc0692245f269b98a3bd4820136/libs/contrib/Decidable/Order.idr#L23), right? That might be worth exploring, although the `ok` thing complicates it a little.

Reply:
Yeah, these `LTE`s are annoying. However, considering zeros destroys most of the properties of factors, since anything is a factor of zero. On the other hand, maybe I should exclude zero only where it really hurts and allow considering factors of zero in general case.

Hmm, indeed! The factor relation is a partial order. That's a thing worth proving. I'll add it.

Reply:
https://en.wikipedia.org/wiki/Order_theory#/media/File:Lattice_of_the_divisibility_of_60.svg